viscy/unet/networks/Unet21D.py

```diff
@@ -1,12 +1,25 @@
-from typing import Sequence, Union
+from typing import Literal, Sequence
 
 import timm
 import torch
-from monai.networks.blocks import ResidualUnit, UnetrUpBlock
+from monai.networks.blocks import ResidualUnit, UpSample
 from monai.networks.blocks.dynunet_block import get_conv_layer
 from torch import nn
 
 
+def _get_convnext_stage(in_channels: int, out_channels: int, depth: int):
+    return timm.models.convnext.ConvNeXtStage(
+        in_chs=in_channels,
+        out_chs=out_channels,
+        stride=1,
+        depth=depth,
+        ls_init_value=None,
+        use_grn=True,
+        norm_layer=timm.layers.LayerNorm2d,
+        norm_layer_cl=timm.layers.LayerNorm,
+    )
+
+
 class Conv21dStem(nn.Module):
```

Comment:
What is the benefit of downsampling the input with `stride = kernel_size`? Why not learn the filters with `stride=1`, and then average pool?

I am worried this may affect the shift invariance of the model by worsening the aliasing effects described [in this paper](https://arxiv.org/pdf/1904.11486.pdf)

Reply:
This is borrowed from [ConvNext](http://arxiv.org/abs/2201.03545). The idea is to patchify the image tokens and learn embeddings with large kernels in deep layers.

In practice modern DCNNs rarely use stride 1 for the first convolution layer, because the high resolution restricts the features that can be learned at each stage. For example ResNet (2015) uses k=7, s=2 followed by 2x2 pooling as its stem. This is especially true for high-resolution 3D input.

Another distinction is that the natural images used in the paper you linked are low resolution, meaning that very small windows (e.g. 3x3) can contain critical texture information in both luminance and chrominance. In microscopy images we train on, voxel sizes are co-designed with diffraction limit and 4x4 patches should not contain more than one distinguishable feature.

Also classification networks suffer from feature space collapse and thus are much more sensitive to this problem than models trained for dense tasks. Since our model generalizes even on a different microscope where the optical transfer is completely different, I don't think there is empirical evidence supporting using stride 1 at the stem layer.

Reply:
Very interesting paper! Pasting the relevant bit of text:
![image](https://github.com/mehta-lab/VisCy/assets/2934183/86c76429-9c64-4283-932f-8b3a98f51316)

The improvement in image classification accuracy with patchification is reported to be 0.1%. That may not be reproducible.

> Another distinction is that the natural images used in the paper you linked are low resolution, meaning that very small windows (e.g. 3x3) can contain critical texture information in both luminance and chrominance.

Our data is typically critically sampled (Nyquist criterion) and we deconvolve to boost high spatial frequencies. Subsampling can lead to aliasing. 

Did the checkerboard artifacts coincide with the use of patchify? 
You may have already done experiments with different strides (levels of patchification). If yes, please point me to them.

Reply:
> The improvement in image classification accuracy with patchification is reported to be 0.1%. That may not be reproducible.

Note that they compared to the ResNet stem, i.e. 7x7 stride 2 convolution followed by max pooling for a total 4x4 downsampling. Both methods use a stem containing one convolution layer to achieve 4x4 down sampling. The difference being that in the later design the downsampling is completely learned. This should actually help *avoid* aliasing.
Stride 1 at the first conv followed by pooling will result in a stem that does not learn the downsampling at all. It is also very expensive given the large 3D input. With the current width, that single layer will take up more than 10% of the FLOPs of the entire network. This is one of the main reasons why the 2.5D network can only learn 16 filters in the first layer with a 48 GB GPU on typical patch sizes. In other words, without the initial aggressive learned projection, the 2.1D network will be more similar to 2.5D in terms of capacity.

> Did the checkerboard artifacts coincide with the use of patchify?

The checkerboard artifacts are due to the upscaling method, and they do not appear when linear upsampling is used. This is a well-studied problem, and I haven't seen any literature linking it to the stride in the first encoder layer.

Reply:
> You may have already done experiments with different strides (levels of patchification). If yes, please point me to them.

I haven't done any experiments since it will require significantly shrinking down the rest of the network. This is not necessarily bad, but will likely require a different overall design. See [HR-Net](https://arxiv.org/pdf/1908.07919v2.pdf) for an example.

At this stage I have been avoiding hyper-parameter tuning unless there is a clear issue it can fix.

Reply:
There is a third option: since the pipeline now supports distributed training, we can scale the model beyond the computational complexity of old 2.5D models and learn both high-resolution feature maps and very deep representations.